backend/etl/processor.js

```diff
@@ -0,0 +1,215 @@
+  }
+
+  _crypto(shop) {
+    return Boolean(shop.listingId)
+  }
+
+  _stripe(shop) {
+    const key = shop.config.stripeBackend
+    return key && key.startsWith('sk_live')
+  }
+
+  _paypal(shop) {
+    return Boolean(shop.config.paypalClientSecret)
+  }
+
+  _uphold(shop) {
+    return Boolean(shop.config.upholdSecret)
+  }
+
+  _manualPayment(shop) {
+    const payments = shop.config.offlinePaymentMethods
+    if (payments && Array.isArray(payments)) {
+      const count = payments.filter((p) => p.disabled).length
+      return count > 0
+    }
+    return false
+  }
+
+  _printful(shop) {
+    return Boolean(shop.config.printful)
+  }
+
+  _sendgrid(shop) {
+    return Boolean(shop.config.sendgridApiKey)
+  }
+
+  _aws(shop) {
+    return Boolean(shop.config.awsAccessSecret)
+  }
+
+  _mailgun(shop) {
+    return Boolean(shop.config.mailgunSmtpPassword)
+  }
+
+  async _published(shop) {
+    const published = await ShopDeployment.findOne({
+      where: { shopId: shop.id, status: ShopDeploymentStatuses.Success }
+    })
+    return Boolean(published)
+  }
+
+  _customDomain(shop) {
+    // TODO(franck): figure out how to extract this information.
+    log.warn(`Shop ${shop.id}: custom domain extraction not implemented yet.`)
+    return null
+  }
+
+  /**
+   * Extracts data for a shop.
+   * @param {models.Shop} shop
+   * @returns {Promise<void>}
+   */
+  async processShop(shop) {
+    const data = {}
+    data.shopId = shop.id
+    data.numProducts = await this._countNumProducts(shop)
+    data.numCollections = await this._countNumCollections(shop)
+    data.numDiscounts = await this._countNumDiscounts(shop)
+    data.numDiscounts = await this._countNumShippings(shop)
+    data.numOrders = await this._countNumOrders(shop)
+    data.crypto = this._crypto(shop)
+    data.stripe = this._stripe(shop)
+    data.paypal = this._paypal(shop)
+    data.uphold = this._uphold(shop)
+    data.manualPayment = this._manualPayment(shop)
+    data.printful = this._printful(shop)
+    data.sendgrid = this._sendgrid(shop)
+    data.aws = this._aws(shop)
+    data.mailgun = this._mailgun(shop)
+    data.published = shop.published
+    data.customDomain = this._customDomain(shop)
+    return data
+  }
+
+  async _decorateShop(shop, netConfig) {
+    // Replace the shop's encrypted config with its decrypted version.
+    shop.config = getConfig(shop.config)
+
+    // Determine if the shop was published. This is used when extracting data
+    // to decide on what source to use: IPFS vs deploy staging area on disk.
+    shop.published = await this._published(shop)
+
+    // Get the shop's data URL.
+    shop.dataUrl = getShopDataUrl(shop, netConfig)
+
+    // Get the shop's deploy dir.
+    shop.outputDir = `${DSHOP_CACHE}/${shop.auth_token}`
+  }
+
+  /**
+   * Main
+   * @returns {Promise<void>}
+   */
+  async run() {
+    log.info('Running ETL job...')
+    const job = await EtlJob.create({ status: EtlJobStatuses.Running })
```

Comment:
Should we make sure a job isn't already running? What happens if we deploy while a job is running?

Reply:
Given the jobs are scheduled only every 24hrs and currently a job should take no more than 5 to 10min max to run, I don't expect there would be any running job when a new one kicks off.

In case the daily job gets interrupted, the failure will get recorded in Sentry.
Then we can simply re-run the job from the Bull admin UI.